convert.c

```diff
@@ -256,6 +257,117 @@ static int will_convert_lf_to_crlf(size_t len, struct text_stat *stats,
 
 }
 
+static struct encoding {
+	const char *name;
+	iconv_t *to_git;
+	iconv_t *to_worktree;
+	struct encoding *next;
+} *encoding, **encoding_tail;
+static const char *default_encoding = "utf-8";
+
+static int encode_to_git(const char *path, const char *src, size_t src_len,
+			 struct strbuf *buf, struct encoding *enc)
+{
+#ifndef NO_ICONV
+	char *dst, *re_src;
+	int dst_len, re_src_len;
+
+	if (!enc || (src && !src_len))
+		return 0;
+
+	/*
+	 * Looks like we got called from "would_convert_to_git()".
+	 * Since we have an encoding we must assume that we would encode
+	 * the content.
+	 */
+	if (!buf && !src) {
+		return 1;
+	}
+
+	if (!enc->to_git || !enc->to_worktree) {
+		enc->to_git = xcalloc(1, sizeof(iconv_t));
```

Comment:
This still looks like a double allocation to me.

First of all, this doesn’t need to be cleared (`xalloc` would also do it), because the pointer is reassigned in the next line.

However, [`iconv_open` allocates the memory for you](http://man7.org/linux/man-pages/man3/iconv_open.3.html) if I read the `man` pages correctly. [Examples](https://www.gnu.org/software/libc/manual/html_node/iconv-Examples.html) also show that allocating this before isn’t necessary.

Reply:
Ah, I see get it now. While `iconv_open` indeed performs an allocation, this is opaque within the `iconv_t` data structure. In other words, your `iconv_t *` is one more level of indirection, which is why manual allocation appears to be necessary.

Reply:
However, `xcalloc` could still be `xalloc`, because the pointer doesn’t need to be nulled (it’s assigned immediately afterward).

I also wonder whether you can’t simply make the `iconv_t *` in the `encoding` struct plain `iconv_t`, assuming that copying `iconv_t` isn’t expensive.

Reply:
Argh. Spot on! Thank you!